### PR TITLE
Fix some issues in DeployDependencies.proj

### PR DESCRIPTION
**Pull request by SamB**
The xplat branch didn't seem to build for me with VS 2015 before, but with these changes it does.

Mostly because NuGet has a much easier time finding all the packages it's supposed to grab if it uses the repositories listed in MSBuild's NuGet.Config rather than relying on ~/.nuget/NuGet.Config or the defaults.

### CODE REVIEW

**Comment by SamB** [comment]
So it looks like something has gone very wrong here, something which is probably connected to the changes that NuGet thinks should be made to `runtime.project.lock.json`.

The only explanation I can think of (given how little I know about any of this) is: perhaps, due to the indeterminate ordering of the timestamps on `runtime.*.json` (which, given the order of their names, most likely manifests in `runtime.project.lock.json` getting a later-or-equal timestamp than `runtime.project.json` from any git operation that updates both files), a stale file has been able to remain undetected for the past few days.  So if anyone has any clues for me, I'd love to hear them!

File: targets/DeployDependencies.proj [hunk 1]
```diff
@@ -25,19 +26,32 @@
...
+         are both under version control, their timestamps are left in an
+         indeterminate order by various operations, such as "git clone",
+         "git pull", and "git checkout".
+    -->
+    <Touch Files="$(RuntimeProjectJson).restored" AlwaysCreate="true" />
+    <!-- We also need to update the timestamp on $(RuntimeProjectLockJson)
+         explicitly; NuGet does not bother to rewrite it when the contents
+         would not change.
+    -->
     <Touch Files="$(RuntimeProjectLockJson)" />
```
Comment (rainersigwald) [inline targets/DeployDependencies.proj]: Why do we still need this?  Can it just be removed from the `Outputs` attribute instead?

**Comment by rainersigwald** [comment]
Changes seem reasonable to me but @dsplaisted should probably take a look.

@SamB if you still have some logs from the problem, can you post snippets here for posterity?  Things were working fine for most of us from VS but with NuGet's machine-wide cache it can be difficult to know when your machine is really clean.

**Comment by rainersigwald** [comment]
Actually, I'm worried about the CI build failures--it seems to all be in this flavor:

    System.IO.FileNotFoundException : Could not load file or assembly 'Microsoft.Win32.Primitives, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. (TaskId:715)
    Stack Trace: (TaskId:715)
          at Microsoft.Build.Shared.FileUtilities.get_CurrentExecutablePath() (TaskId:715)
       XMake_Tests.cs(873,0): at Microsoft.Build.UnitTests.XMakeAppTests..ctor() (TaskId:715)
     TestProcessProjectSwitch [FAIL] (TaskId:715)

Which seems likely to be related to a file not getting copied from the `runtime` list . . .

**Comment by SamB** [comment]
(Or maybe `xplat` builds just always fail on the bot?)

**Comment by rainersigwald** [comment]
xplat builds should be passing now (and seem to be for other PRs). That's what I spent most of the last couple of weeks on. . .

File: targets/DeployDependencies.proj (hunk 1 above)
Comment (SamB) [inline targets/DeployDependencies.proj]: Yeah, it can probably just be removed.

**Comment by SamB** [comment]
@rainersigwald Perhaps it would be good to make provision for avoiding the machine-wide cache, to make it easier to test the clean build case? In other words, set up a way, like a build parameter or environment variable, to arrange that NuGet's -NoCache option be passed down to all relevant nuget commands (which seem to be just install and restore).

**Comment by SamB** [comment]
Well, it mostly works for me; some tests fail, but I didn't see anything too bad, except that xunit.runner.visualstudio complains that it can't find xunit.execution.desktop.dll and gives up on each test assembly, first for discovery and then for execution.  I would suggest that there might be too much stuff in the buildbot's msbuild/packages directory, except doesn't the buildbot clone fresh each time?

**Comment by SamB** [comment]
Okay, I took out that gratuitous `$(RuntimeProjectLockJson)` in the outputs, moved the stamp file to be inside the packages directory, and found out that the `ResolveNuGetPackageAssets` task was trying to find packages in `~/.nuget/packages` (which I have now fixed).

**Comment by SamB** [comment]
Hmm, but it still fails to load one of the test assemblies in that ominous way :-(.  Guess there's more digging to do here ...

**Comment by SamB** [comment]
And it turns out that I do not get any such ominous when I run `RebuildWithLocalMSBuild.cmd`, which makes it difficult for me to investigate it.

It did fail the first time I ran it, though, saying:
<pre>
  Build binary target directory: 'C:\Users\Sam\Source\Repos\msbuild\bin\Windows
  _NT\Debug-NetCore\'
  Restoring NuGet packages...
C:\Users\Sam\Source\Repos\msbuild\src\Framework\Microsoft.Build.Framework.cspro
j : error MSB4057: The target "_NuGet_GetProjectsReferencingProjectJsonInternal
" does not exist in the project. [C:\Users\Sam\Source\Repos\msbuild\build.proj]
  Process must exit before requested information can be determined.
C:\Users\Sam\Source\Repos\msbuild\build.proj(24,5): error MSB3073: The command
"C:\Users\Sam\Source\Repos\msbuild\packages\NuGet.exe restore "C:\Users\Sam\Sou
rce\Repos\msbuild\src\MSBuild.sln"" exited with code 1.

    0 Warning(s)
    2 Error(s)
</pre>
(The first message actually appeared 13 times.)

The second time through, though, the main things I get are various types of warnings about conflicts between assemblies and some stuff about skipped/failed tests.  And the error from the test runner returning a non-zero value, of course.  I guess the only things left for me to try are:

1. Deleting `packages` and `bin/Windows_NT/Debug-NetCore` and running `RebuildWithLocalMSBuild.cmd` again
2. The same thing, only with `~/.nuget/packages` moved aside

**Comment by rainersigwald** [comment]
I finally got some time to set up a nice clean (checkpointed) VM and run a build, and I reproduced the `Microsoft.Win32.Primitives` problems.  I'm digging in.

**Comment by SamB** [comment]
@rainersigwald Oh, I probably should have mentioned that I managed to reproduced that too, but got lost trying to figure out what thing had failed to mention that it depended on that package.

**Comment by rainersigwald** [comment]
@dotnet-bot test this please

**Comment by rainersigwald** [comment]
@SamB After further chasing, it turned out that a bug in NuGet restore caused files to be missing from the list of files in the runtime on clean machines.  I worked around it for now with a double-restore (which corrupts and then fixes `runtime.project.lock.json`), and it got fixed with https://github.com/NuGet/NuGet.Client/commit/42cbee2a8aca0f448d09734a753245a60dedaa2f.  We can remove the workaround when we adopt NuGet 3.3 after its release.

**Comment by rainersigwald** [comment]
@dotnet-bot test this please

File: .gitignore [hunk 2]
```diff
@@ -225,3 +225,7 @@ ApiPort/
 *.xlsx
 ref/
 Differences.txt
+# "stamp" files, dummy outputs that exist only to hold timestamps
```
Comment (rainersigwald) [inline .gitignore]: I think we tend to call these "semaphores", like `BuildTools.semaphore`.

File: targets/DeployDependencies.proj [hunk 3]
```diff
@@ -25,20 +26,32 @@
...
-    <!-- It appears that NuGet doesn't re-write the project.lock.json file if the contents would be the same. This can mess up incremental builds
-        if you make a non-significant change to project.json.  To avoid that, update the timestamp of the lock file here. -->
-    <Touch Files="$(RuntimeProjectLockJson)" />
+    <!-- This is just a "stamp" file, serving no purpose besides indicating
+         that the target has been run since $(RuntimeProjectJson) was last
+         updated.  The reason we can't use $(RuntimeProjectLockJson) for this
+         is that, since $(RuntimeProjectJson) and $(RuntimeProjectLockJson)
+         are both under version control, their timestamps are left in an
+         indeterminate order by various operations, such as "git clone",
+         "git pull", and "git checkout".
```
Comment (rainersigwald) [inline targets/DeployDependencies.proj]: I think this is a solid, convincing reason to use a separate file as the semaphore.

File: targets/DeployDependencies.proj [hunk 4]
```diff
@@ -25,20 +26,32 @@
   <PropertyGroup>
     <RuntimeProjectJson>$(MSBuildThisFileDirectory)\runtime.project.json</RuntimeProjectJson>
     <RuntimeProjectLockJson>$(MSBuildThisFileDirectory)\runtime.project.lock.json</RuntimeProjectLockJson>
-    <RestoreRuntimePackagesCommand>"$(NuGetToolPath)" restore $(RuntimeProjectJson)</RestoreRuntimePackagesCommand>
+    <RestoreRuntimePackagesCommand>$(NugetRestoreCommand) $(NuGetPackageInstallOptions) $(RuntimeProjectJson)</RestoreRuntimePackagesCommand>
```
Comment (rainersigwald) [inline targets/DeployDependencies.proj]: It *shouldn't* be required to restore these packages to our local packages directory.  The idea is to restore them to your machine/user NuGet cache, then use `ResolveNuGetPackageAssets` to copy them locally.  I believe that this should be robust now after my change to run the restore twice to work around the NuGet bug.

**Comment by rainersigwald** [comment]
I have no idea why this change would cause such test flakiness.  But regardless, I don't think moving runtime restore to the local packages folder is necessary.  Can you try the scenarios that were causing you trouble again after 66dffed1c142c838eaff8f0c1b42849757bde9e2?

The `.gitignore` and external-semaphore-for-restore changes are good; I would want to take them.  And obviously I want a robust clone-and-build experience, but I hope the changes there aren't required.

**Comment by SamB** [comment]
@rainersigwald: I can't help but find the idea of actually deleting my nuget cache somewhat scary; any tips on a way to simulate this?

**Comment by rainersigwald** [comment]
That's fair enough, @SamB!  Let's assume that it's working (it has done OK in my clean VMs) and do this:

- [ ] Remove the local-packages-directory change (shouldn't be required, hurts machine-cache efficiency).
- [ ] Rename the file from "stamp" to "sempahore" for consistency.
- [ ] Rebase onto current master.

Those seem like good changes regardless, and should improve robustness as well.  If we get another report of flakiness we can investigate it then.

Sound ok?

**Comment by SamB** [comment]
@rainersigwald: Those sound reasonable. FYI, I used "stamp" because, where I come from, a semaphore is more-or-less a sort of lock, so it took me a while to work out that you guys use it for files which don't store any meaningful information aside from their timestamp.

**Comment by AndyGerlicher** [comment]
This PR hasn't been looked at in a while. Sorry to keep it open so long and thanks for the contribution. These files have diverged so much I'm not sure if this is still relevant or not. Please reopen / rebase if you think it's still needed.

**Pull request closed without merging**